### PR TITLE
2243: Unit tests card extensions

### PR DESCRIPTION
**Pull request by f1sh1918**
### Short Description

In this pr i added the missing unit tests for card extensions. I spared out some fromString and toString methods if the actual state type is a string

### Proposed Changes

- add missing unit tests for extensions
- moved one unit test to the correct folder

Note: Tests for `fullName` and `expirationDate` are missing, since these are not extensions and have to be tested separately. I will create a ticket for that.

### Side Effects

- none

### Testing

N/A

### Resolved Issues

Fixes: #2243

### CODE REVIEW

**Comment by seluianova** [comment]
I think would be nice to cover all corner cases where we found issues here https://github.com/digitalfabrik/entitlementcard/pull/2201

**Comment by f1sh1918** [comment]
> I think would be nice to cover all corner cases where we found issues during testing of #2201

yes i moved it to draft

**Comment by f1sh1918** [comment]
hey @steffenkleinle or someone else that has good knowledge in unit tests and react, i could need some help here. I have the problem that in unit tests my error messages won't be recreated, when extension state changes
1. Initial state value of an extension is set
2. I click the clear button and do blur
3. If i just show an error message it works, but if i have to use the state of the extension it still uses the initialValue and not the cleared `null` value to show the correct error message.
The display of the error message in gui works as expected.

**Review by bahaaTuffaha — APPROVED** [review]
Nice work ✨
There is one issue and the others are suggestions.

File: administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx [hunk 1]
```diff
@@ -70,6 +118,49 @@ describe('BirthdayExtension', () => {
         )
       ).toBeTruthy()
     })
+
+    it('should not show an hint if provided birthday is underage for nuernberg', () => {
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx:122]:
```suggestion
    it('should not show a hint if provided birthday is underage for nuernberg', () => {
```

File: administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx [hunk 2]
```diff
@@ -15,46 +17,92 @@ const wrapper = ({ children }: { children: ReactNode }) => (
...
+        { wrapper }
+      )
       expect(getByText('Das Geburtsdatum darf nicht in der Zukunft liegen.')).toBeTruthy()
     })
 
-    it('should show an hint if provided birthday is underage', () => {
-      const setValue = jest.fn()
-      const tomorrow = new Date()
-      tomorrow.setDate(tomorrow.getDate() + 1)
+    it('should show an hint if provided birthday is underage for koblenz', () => {
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx:105]:
```suggestion
    it('should show a hint if provided birthday is underage for koblenz', () => {
```

File: administration/src/cards/extensions/__tests__/KoblenzReferenceNumberExtension.test.tsx [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+          isValid={false}
+          showRequired
+        />
+      )
+      const input = getByDisplayValue(initialValue)
+      fireEvent.blur(input)
+      expect(getByText('Das Aktenzeichen enthält ungültige Sonderzeichen.')).toBeTruthy()
+    })
+
+    it('should show error for invalid length', () => {
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/KoblenzReferenceNumberExtension.test.tsx:77]: I think would be nice to have 2 tests here: length 3 and 16

File: administration/src/cards/extensions/__tests__/StartDayExtension.test.tsx [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+      )
+      const datePicker = getByDisplayValue(startDay.format())
+      fireEvent.blur(datePicker)
+      expect(
+        getByText(`Das Startdatum darf nicht weiter als ${minStartDay.format()} in der Vergangenheit liegen.`)
+      ).toBeTruthy()
+    })
+
+    it('should show error message when startDay is too far in the future and touched', () => {
+      const startDay = new PlainDate(2200, 1, 1)
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/StartDayExtension.test.tsx:69]: same here.
I would recommend to test max start date + 1 as invalid value and max start date as valid value

File: administration/src/cards/extensions/__tests__/StartDayExtension.test.tsx [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+      })
+
+      it('should result in minStartDay if no startDay is provided', () => {
+        expect(StartDayExtension.getProtobufData({ startDay: null })).toEqual({
+          extensionStartDay: { startDay: 18424 },
+        })
+      })
+    })
+
+    describe('isValid', () => {
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/StartDayExtension.test.tsx:123]: ❓ maybe these tests can be combined with the ones where we check for error messages / no error messages ?
otherwise these tests should be duplicated, because we need to check that the field is invalid, in the same cases where we are checking if an error message is displayed

File: administration/src/cards/extensions/components/__tests__/CardTextField.test.tsx [hunk 6]
```diff
@@ -19,7 +19,7 @@ describe('CardTextField', () => {
     const { getByLabelText } = render(<CardTextField {...defaultProps} />)
 
     const input = getByLabelText('Test Label')
-    expect(input).toBeInTheDocument()
+    expect(input).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (f1sh1918) [reply administration/src/cards/extensions/components/__tests__/CardTextField.test.tsx:22]: normale you use `toBeInTheDocument` if you check the whole DOM f.e. `screen.getByXY`
This is not needed here

File: administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx [hunk 7]
```diff
@@ -15,46 +17,92 @@ const wrapper = ({ children }: { children: ReactNode }) => (
...
         <BirthdayExtension.Component
           showRequired
           setValue={setValue}
           isValid={false}
-          value={{ birthday: PlainDate.fromLocalDate(tomorrow) }}
+          value={{ birthday: new PlainDate(1899, 1, 1) }}
         />,
         { wrapper }
       )
+      expect(getByText('Das Geburtsdatum darf nicht vor dem 01.01.1900 liegen.')).toBeTruthy()
```
[the remark this answers is not in the document]
Comment (seluianova) [reply administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx:93]: But checking isValid() and checking to display validation error are 2 different checks. And we had an issue where a field was marked as invalid, but no validation error was displayed.
I think these checks should be either combined (one test for both checks) or duplicated.

[the remark this answers is not in the document]
Comment (f1sh1918) [reply administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx:99]: i combined this test with the last commit

File: administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx [hunk 8]
```diff
@@ -15,60 +17,203 @@ const wrapper = ({ children }: { children: ReactNode }) => (
...
+    })
+
+    it('should not show error if provided birthday is today', () => {
+      const { queryByText } = renderWithTranslation(
+        <BirthdayExtension.Component showRequired setValue={setValue} isValid={false} value={{ birthday: today }} />,
+        { wrapper }
+      )
+      expect(BirthdayExtension.isValid({ birthday: today })).toBeTruthy()
+      expect(queryByText('Das Geburtsdatum darf nicht in der Zukunft liegen.')).toBeNull()
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx:108]: 💭 I think it's better to check that there is no errors in general in negative tests, rather than “absence of one particular error”. 
For example, by the name or id of the component containing the errors

File: administration/src/cards/extensions/__tests__/KoblenzReferenceNumberExtension.test.tsx [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+          showRequired={false}
+        />
+      )
+
+      const clearButton = getByRole('button')
+      fireEvent.click(clearButton)
+      expect(mockSetValue).toHaveBeenCalledWith({ koblenzReferenceNumber: '' })
+    })
+
+    describe('isValid', () => {
```
Comment (seluianova) [inline administration/src/cards/extensions/__tests__/KoblenzReferenceNumberExtension.test.tsx:122]: can be merged with component tests (same reason as discussed in birthday extension)

**Review by seluianova — APPROVED** [review]
a few minor suggestions, but the coverage is already very good in my opinion 👍

File: administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx (hunk 8 above)
Comment (f1sh1918) [reply administration/src/cards/extensions/__tests__/BirthdayExtension.test.tsx:108]: okay i agree on this if different errors can be expected and you can not test against one particular error that it will not be shown.
I adjusted these cases.
But for the birthday hint, you have one particular text to check if its there or not, so this test is more explicit and i will keep it there
